pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java

```diff
@@ -0,0 +1,656 @@
+                try {
+                    for (Map.Entry<String, String> entry : bundlesToUnload.entrySet()) {
+                        final String bundle = entry.getKey();
+                        final String broker = entry.getValue();
+                        adminCache.get(broker).namespaces().unloadNamespaceBundle(
+                                getNamespaceNameFromBundleName(bundle), getBundleRangeFromBundleName(bundle));
+                    }
+                } catch (Exception e) {
+                    log.warn("Error when trying to perform load shedding: {}", e);
+                }
+                return;
+            }
+        }
+    }
+
+    /**
+     * As the leader broker, attempt to automatically detect and split hot namespace bundles.
+     */
+    @Override
+    public void doNamespaceBundleSplit() {
+        // TODO?
+    }
+
+    @Override
+    public String getBrokerRoot() {
+        return LOADBALANCE_BROKERS_ROOT;
+    }
+
+    /**
+     * When the broker data ZooKeeper nodes are updated, update the broker data map.
+     */
+    @Override
+    public void onUpdate(final String path, final LocalBrokerData data, final Stat stat) {
+        scheduler.submit(this::updateAll);
+    }
+
+    /**
+     * As the leader broker, find a suitable broker for the assignment of the given bundle.
+     * 
+     * @param serviceUnit
+     *            ServiceUnitId for the bundle.
+     * @return The name of the selected broker, as it appears on ZooKeeper.
+     */
+    @Override
+    public synchronized String selectBrokerForAssignment(final ServiceUnitId serviceUnit) {
+        // ?: Is it too inefficient to make this synchronized? If so, it may be
+        // a good idea to use weighted random
+        // or atomic data.
+
+        final String bundle = serviceUnit.toString();
+        if (preallocatedBundleToBroker.containsKey(bundle)) {
+            // If the given bundle is already in preallocated, return the
+            // selected broker.
+            return preallocatedBundleToBroker.get(bundle);
+        }
+        final BundleData data = loadData.getBundleData().computeIfAbsent(bundle, key -> getBundleDataOrDefault(bundle));
+        brokerCandidateCache.clear();
+        brokerCandidateCache.addAll(loadData.getBrokerData().keySet());
+        policyFilter(serviceUnit);
+
+        // Use the filter pipeline to finalize broker candidates.
+        for (BrokerFilter filter : filterPipeline) {
+            filter.filter(brokerCandidateCache, data, loadData, conf);
+        }
+        final String broker = placementStrategy.selectBroker(brokerCandidateCache, data, loadData, conf);
+
+        // Add new bundle to preallocated.
+        loadData.getBrokerData().get(broker).getPreallocatedBundleData().put(bundle, data);
+        preallocatedBundleToBroker.put(bundle, broker);
+        return broker;
+    }
+
+    private void policyFilter(final ServiceUnitId serviceUnit) {
```

Comment:
is the same logic compare to `SimpleLoadManagerImpl.getFinalCandidates()` to filter out candidate if yes then should we put it into some common place to avoid for any new loadManager?